### PR TITLE
Fix await parsing in modules

### LINKED ISSUE #4189: Invalid memory access (segfault) in opfunc_resume_executable_object in opcodes.c
###### JerryScript revision
```
The master branch or git hash: c013fad
```

###### Build platform
```
Ubuntu 18.04.4 LTS  Linux 4.15.0-74-generic x86_64
```

###### Build steps
```
Build with ASAN enabled:
python tools/build.py  --compile-flag=-fsanitize=address --compile-flag=-g --strip=OFF
```
###### Test case
```
1 ; export default await 1
```

###### Output
```
ASAN:DEADLYSIGNAL
=================================================================
==7178==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000004 (pc 0x55555558d55f bp 0x555555855110 sp 0x7fffffffde50 T0)
==7178==The signal is caused by a READ memory access.
==7178==Hint: address points to the zero page.
    #0 0x55555558d55e in opfunc_resume_executable_object /home/user/Product/jerryscript/master/src/jerry-core/vm/opcodes.c:789
    #1 0x555555599dc4 in ecma_process_promise_async_reaction_job /home/user/Product/jerryscript/master/src/jerry-core/ecma/operations/ecma-jobqueue.c:319
    #2 0x555555599dc4 in ecma_process_all_enqueued_jobs /home/user/Product/jerryscript/master/src/jerry-core/ecma/operations/ecma-jobqueue.c:529
    #3 0x55555556cb80 in jerry_run_all_enqueued_jobs /home/user/Product/jerryscript/master/src/jerry-core/api/jerry.c:626
    #4 0x55555556cb80 in main /home/user/Product/jerryscript/master/src/jerry-main/main-unix.c:286
    #5 0x7ffff6a4ab96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)
    #6 0x55555556d5d9 in _start (/home/user/Product/jerryscript/master/src/build/bin/jerry+0x195d9)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV /home/user/Product/jerryscript/master/src/jerry-core/vm/opcodes.c:789 in opfunc_resume_executable_object
==7178==ABORTING
```

### PR DESCRIPTION
**Pull request by pmarkee**
Fixes #4189

JerryScript-DCO-1.0-Signed-off-by: Peter Marki marpeter@inf.u-szeged.hu

### CODE REVIEW

**Comment by pmarkee** [comment]
Note: the behaviour is made to be compatible with Node.js.

File: tests/jerry/es.next/function-await1.js [hunk 1]
```diff
@@ -51,11 +55,7 @@ async (a) => {
 }
 
 (a) => {
-  await
   async (a) => await a
-  await
-  async (a) => await a
-  a\u0077ait
 }
```
Comment (zherczeg) [inline tests/jerry/es.next/function-await1.js:59]: Why these needs to be changed? FF and Chrome accepts them.

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: tests/jerry/es.next/function-await1.js (hunk 1 above)
Comment (pmarkee) [reply tests/jerry/es.next/function-await1.js:59]: I followed node's behaviour, and node only allows `await` inside async functions. What I overlooked is that node will only error if the `await` actually has an argument, but if `await` stands by itself then it is an identifier, for example:
```
function f(a) {
  await; // OK in node (sees it as an identifier), SyntaxError in jerry
  await a; // SyntaxError in node and jerry
}
```
But V8 and FF have implemented the top level await proposal, which we have not, so we have to deal with this somehow. I think it's okay to now allow await anywhere outside of async until we implement the proposal too.

**Review by zherczeg — APPROVED** [review]
LGTM

File: jerry-core/parser/js/js-parser.c [hunk 2]
```diff
@@ -3222,6 +3216,12 @@ parser_parse_script (const uint8_t *arg_list_p, /**< function argument list */
                                            resource_name,
                                            line_str_val,
                                            col_str_val);
+#if ENABLED (JERRY_MODULE_SYSTEM)
+    if (JERRY_CONTEXT (module_top_context_p) != NULL)
+    {
+      ecma_module_cleanup ();
+    }
+#endif
```
Comment (rerobika) [inline jerry-core/parser/js/js-parser.c:3224]: Please add the missing `ENABLED (JERRY_MODULE_SYSTEM)` comment

File: jerry-core/parser/js/js-parser-expr.c [hunk 3]
```diff
@@ -1907,6 +1907,15 @@ parser_parse_unary_expression (parser_context_t *context_p, /**< context */
 #if ENABLED (JERRY_ESNEXT)
       case LEXER_KEYW_AWAIT:
       {
+#if ENABLED (JERRY_MODULE_SYSTEM)
+        if ((context_p->global_status_flags & ECMA_PARSE_MODULE)
+            && !(context_p->status_flags & PARSER_IS_ASYNC_FUNCTION))
+        {
+          parser_raise_error (context_p, PARSER_ERR_AWAIT_NOT_ALLOWED);
+          break;
```
Comment (rerobika) [inline jerry-core/parser/js/js-parser-expr.c:1915]: I think this break is not necessary since `parser_raise_error` uses longjump.

**Review by rerobika — CHANGES_REQUESTED** [review]

**Comment by pmarkee** [comment]
@rerobika Done.

**Review by rerobika — APPROVED** [review]
LGTM

**Pull request merged**